File: Other/Review/Phase 9/Review2/Part1/heap-sort/heap-sort.ts
function heapify(nums: number[], length: number, current_index: number) {
  let max_index = current_index;
  let left_index = 2 * current_index + 1;
  let right_index = 2 * current_index + 2;

  if (left_index < length && nums[max_index] < nums[left_index]) {
    max_index = left_index;
  }
  if (right_index < length && nums[max_index] < nums[right_index]) {
    max_index = right_index;
  }

  if (current_index != max_index) {
    const temp = nums[current_index];
    nums[current_index] = nums[max_index];
    nums[max_index] = temp;
    heapify(nums, length, max_index);
  }
}

export function heapSort(nums: number[]) {
  for (let i = Math.floor(nums.length / 2) - 1; i >= 0; i--) {
    heapify(nums, nums.length, i);
  }

  for (let i = nums.length - 1; i >= 0; i--) {
    const temp = nums[i];
    nums[i] = nums[0];
    nums[0] = temp;

    heapify(nums, i, 0);
  }
}

// test
const array = [2, 0, 2, 1, 1, 0, -3, -4];
heapSort(array);
console.log(array);

File: Data Structure & Algorithm/Review/Stage 6/Review 6/Part1/heap-sort/heap-sort.go
package main

import (
	"fmt"
)

func heapify(nums []int, length int, currentIndex int) {
	maxIndex := currentIndex
	leftIndex := 2*currentIndex + 1
	rightIndex := 2*currentIndex + 2

	if leftIndex < length && nums[maxIndex] < nums[leftIndex] {
		maxIndex = leftIndex
	}
	if rightIndex < length && nums[maxIndex] < nums[rightIndex] {
		maxIndex = rightIndex
	}

	if currentIndex != maxIndex {
		temp := nums[currentIndex]
		nums[currentIndex] = nums[maxIndex]
		nums[maxIndex] = temp
		heapify(nums, length, maxIndex)
	}
}

func heapSort(nums []int) {
	for i := len(nums)/2 - 1; i >= 0; i-- {
		heapify(nums, len(nums), i)
	}

	for i := len(nums) - 1; i >= 0; i-- {
		temp := nums[0]
		nums[0] = nums[i]
		nums[i] = temp
		heapify(nums, i, 0)
	}
}

// test
func main() {
	array := []int{2, 0, 2, 1, 1, 0, -3, -4}
	heapSort(array)
	fmt.Println(array)
}
